hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/LruBlockCache.java

```diff
@@ -947,7 +1031,94 @@ public void run() {
+        if (stopTime - startTime <= 1000 * 10 - 1) {
+          mbFreedSum += bytesFreed/1024/1024;
+        } else {
+          // Here we have to calc what situation we have got.
+          // We have the limit "hbase.lru.cache.heavy.eviction.bytes.size.limit"
+          // and can calculte overhead on it.
+          // We will use this information to decide, 
+          // how to change percent of caching blocks.
+          freedDataOverheadPercent =
+            (int) (mbFreedSum * 100 / cache.heavyEvictionMbSizeLimit) - 100;
+          if (freedDataOverheadPercent > 0) {
+            // Now we are in the situation when we are above the limit
+            // But maybe we are going to ignore it because it will end quite soon
+            heavyEvictionCount++;
+            if (heavyEvictionCount > cache.heavyEvictionCountLimit) {
+              // It is going for a long time and we have to reduce of caching
+              // blocks now. So we calculate here how many blocks we want to skip.
+              // It depends on: 
+              // 1. Overhead - if overhead is big we could more aggressive
+              // reducing amount of caching blocks.
+              // 2. How fast we want to get the result. If we know that our 
+              // heavy reading for a long time, we don't want to wait and can 
+              // increase the coefficient and get good performance quite soon.
+              // But if we don't sure we can do it slowly and it could prevent 
+              // premature exit from this mode. So, when the coefficient is 
+              // higher we can get better performance when heavy reading is stable.
+              // But when reading is changing we can adjust to it and set 
+              // the coefficient to lower value.
+              int ch = (int) (freedDataOverheadPercent * cache.heavyEvictionOverheadCoefficient);
+              // But practice shows that 15% of reducing is quite enough.
+              // We are not greedy (it could lead to premature exit).
+              ch = ch > 15 ? 15 : ch;
+              ch = ch < 0 ? 0 : ch; // I think it will never happen but check for sure
+              // So this is the key point, here we are reducing % of caching blocks
+              cache.cacheDataBlockPercent -= ch;
+              // If we go down too deep we have to stop here, 1% any way should be.
+              cache.cacheDataBlockPercent =
+                cache.cacheDataBlockPercent < 1 ? 1 : cache.cacheDataBlockPercent;
+            }
+          } else {
+            // Well, we have got overshooting. 
+            // Mayby it is just short-term fluctuation and we can stay in this mode.
+            // It help avoid permature exit during short-term fluctuation.
+            // If overshooting less than 90%, we will try to increase the percent of 
+            // caching blocks and hope it is enough.
+            if (mbFreedSum >= cache.heavyEvictionMbSizeLimit * 0.1) {
```

Comment:
How did you come up with 0.1? This is critical because this branch controls how the cache recovers to a normal state after heavy eviction is over.

Reply:
If reading fall too much it looks like it over and somewhere  we have to give up (exit from this mode). I thought 90% is quite enough, but it could be 95% and 80%. We have an option - can give control all of it via new parameters. On the  other hand I affraid of users could be embarased and spend a lot of time try to find out what to do.